### DISCUSSION TITLE
"Error updating branch: update failure" with regex manager and docker datasource

### DISCUSSION THREAD

**Discussion by 9numbernine9** [renovatebot/renovate#16720] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Hello! :wave: 

I'm trying to use the `regex` manager to update a Dockerfile based on the [Advanced Capture](https://docs.renovatebot.com/modules/manager/regex/#advanced-capture) section in the documentation. When using the hosted Renovate application, the following message appears in the log:
```
WARN: Error updating branch: update failure(branch="renovate/rust-1.x")
```
The warning message is a little short on details. Look earlier in the logs, it _seems_ like it should do the right thing:
```
{
  "config": {
    "regex": [
      {
        "packageFile": "Dockerfile",
        "deps": [
          {
            "depName": "rust",
            "currentValue": "1.62.0",
            "datasource": "docker",
            "versioning": "semver",
            "replaceString": "datasource=docker depName=rust versioning=semver\nARG RUST_VERSION=1.62.0\n",
            "depIndex": 0,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "1.62.1",
                "newValue": "1.62.1",
                "newMajor": 1,
                "newMinor": 62,
                "updateType": "patch",
                "branchName": "renovate/rust-1.x"
              }
            ],
            "warnings": [],
            "currentVersion": "1.62.0",
            "isSingleVersion": true,
            "fixedVersion": "1.62.0"
          }
        ],
        "matchStrings": [
          "datasource=(?<datasource>.*?) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?\\sARG .*?_VERSION=(?<currentValue>.*)\\s"
        ],
        "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}",
        "autoReplaceStringTemplate": "{{newVersion}}"
      }
    ]
  }
}
```

I have [created a sample repository here](https://github.com/9numbernine9/renovate_debug) that replicates the problem in a minimal way. 

Other notes:
- This _seems_ similar to [this issue here](https://github.com/renovatebot/renovate/issues/15318) but that one is related to Helm resources so it may or may not be relevant.
- I originally encountered this issue with a self-hosted Renovate instance pointing to a repository in Bitbucket Cloud. The example repo contains the relevant portion extracted from that repo.

### Relevant debug logs

<details><summary>Logs</summary>

```
INFO: Dependency extraction complete
{
  "baseBranch": "main",
  "stats": {
    "managers": {
      "regex": {
        "fileCount": 1,
        "depCount": 1
      }
    },
    "total": {
      "fileCount": 1,
      "depCount": 1
    }
  }
}
DEBUG: PackageFiles.add() - Package file saved for branch
{
  "baseBranch": "main"
}
DEBUG: Package releases lookups complete
{
  "baseBranch": "main"
}
DEBUG: branchifyUpgrades
DEBUG: 1 flattened updates found: rust
DEBUG: Returning 1 branch(es)
DEBUG: config.repoIsOnboarded=true
DEBUG: packageFiles with updates
{
  "config": {
    "regex": [
      {
        "packageFile": "Dockerfile",
        "deps": [
          {
            "depName": "rust",
            "currentValue": "1.62.0",
            "datasource": "docker",
            "versioning": "semver",
            "replaceString": "datasource=docker depName=rust versioning=semver\nARG RUST_VERSION=1.62.0\n",
            "depIndex": 0,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "1.62.1",
                "newValue": "1.62.1",
                "newMajor": 1,
                "newMinor": 62,
                "updateType": "patch",
                "branchName": "renovate/rust-1.x"
              }
            ],
            "warnings": [],
            "currentVersion": "1.62.0",
            "isSingleVersion": true,
            "fixedVersion": "1.62.0"
          }
        ],
        "matchStrings": [
          "datasource=(?<datasource>.*?) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?\\sARG .*?_VERSION=(?<currentValue>.*)\\s"
        ],
        "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}",
        "autoReplaceStringTemplate": "{{newVersion}}"
      }
    ]
  }
}
DEBUG: processRepo()
DEBUG: Processing 1 branch: renovate/rust-1.x
DEBUG: Calculated maximum PRs remaining this run
{
  "prsRemaining": 99
}
DEBUG: PullRequests limit = 99
DEBUG: Calculated maximum branches remaining this run
{
  "branchesRemaining": 99
}
DEBUG: Branches limit = 99
DEBUG: getBranchPr(renovate/rust-1.x)(branch="renovate/rust-1.x")
DEBUG: findPr(renovate/rust-1.x, undefined, open)(branch="renovate/rust-1.x")
DEBUG: getPrList success(branch="renovate/rust-1.x")
{
  "pullsTotal": 0,
  "requestsTotal": 1,
  "apiQuotaAffected": true
}
DEBUG: findPr(renovate/rust-1.x, undefined, closed)(branch="renovate/rust-1.x")
DEBUG: branchExists=false(branch="renovate/rust-1.x")
DEBUG: dependencyDashboardCheck=undefined(branch="renovate/rust-1.x")
DEBUG: recreateClosed is false(branch="renovate/rust-1.x")
DEBUG: findPr(renovate/rust-1.x, Update dependency rust to v1.62.1, !open)(branch="renovate/rust-1.x")
DEBUG: prAlreadyExisted=false(branch="renovate/rust-1.x")
DEBUG: Checking schedule(at any time, null)(branch="renovate/rust-1.x")
DEBUG: No schedule defined(branch="renovate/rust-1.x")
DEBUG: Setting current branch to main(branch="renovate/rust-1.x")
DEBUG: latest commit(branch="renovate/rust-1.x")
{
  "branchName": "main",
  "latestCommitDate": "2022-07-22T12:47:40-04:00"
}
DEBUG: Branch needs creating(branch="renovate/rust-1.x")
DEBUG: Using reuseExistingBranch: false(branch="renovate/rust-1.x")
DEBUG: manager.getUpdatedPackageFiles() reuseExistinbranch=false(branch="renovate/rust-1.x")
DEBUG: Starting search at index 12(packageFile="Dockerfile", branch="renovate/rust-1.x")
{
  "depName": "rust"
}
DEBUG: Found match at index 12(packageFile="Dockerfile", branch="renovate/rust-1.x")
{
  "depName": "rust"
}
DEBUG: Could not extract package file(packageFile="Dockerfile", branch="renovate/rust-1.x")
{
  "manager": "regex"
}
WARN: Error updating branch: update failure(branch="renovate/rust-1.x")
DEBUG: Closing Dependency Dashboard
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by Churro** [comment, marked as answer]
Remove `"autoReplaceStringTemplate": "{{newVersion}}"` from your `renovate.json` and it will work as expected 👍

**Comment by 9numbernine9** [comment]
Thanks @Churro , that does seem to work! 👍 

Would it be possible to add additional debug/warning information in this scenario? I don't think I would've tried your solution in a million years based on the warning message. I thought for certain it was a permissions issue!

**Comment by Churro** [comment]
The [page you referenced](https://docs.renovatebot.com/modules/manager/regex/#advanced-capture) doesn't include `autoReplaceStringTemplate`, which is why nobody should encounter this accidentially. The docs for [autoreplacestringtemplate](https://docs.renovatebot.com/configuration-options/#autoreplacestringtemplate) clearly describes its purpose: _Allows overwriting how the matched string is replaced_

A debug message couldn't protect against wrong or unintended usage of config options...

**Comment by 9numbernine9** [comment]
Thanks @Churro . I'm not even sure how the `autoReplaceStringTemplate` got in there - that was more than likely debugging stuff left over from a different issue that I was trying to resolve. :sweat_smile: Everything works perfectly with the change that you suggested.

The reason that I mentioned the debugging/logging thing previously is that (IMHO) I find the `Error updating branch: update failure` to be a little misleading or at least lacking some detail. Based on your description of the issue it seems like the root cause was that Renovate was (correctly) generating a bad replacement string based on my misconfiguration, but without your explanation `Error updating branch` seemed more like a Git or Bitbucket issue. A debug or trace message with a bit more detail would have pointed me in the right direction sooner. :smile:
